src/plugins/intel_gpu/src/plugin/transformations/disable_f16_comp_for_periodic_funcs.cpp

```diff
@@ -0,0 +1,120 @@
+#include "openvino/op/strided_slice.hpp"
+#include "openvino/op/tile.hpp"
+#include "openvino/op/identity.hpp"
+#include "openvino/op/pad.hpp"
+#include "openvino/op/util/gather_base.hpp"
+#include "openvino/op/util/gather_nd_base.hpp"
+#include "openvino/op/squeeze.hpp"
+#include "openvino/op/unsqueeze.hpp"
+#include "disable_f16_comp_for_periodic_funcs.hpp"
+
+#include <unordered_map>
+#include <typeindex>
+
+static bool is_non_value_modifying_node(const std::shared_ptr<ov::Node>& node) {
+    static const std::unordered_map<std::string, bool> node_behavior_registry = {
+        {ov::op::v1::Reshape::get_type_info_static().name, true},
+        {ov::op::v1::Transpose::get_type_info_static().name, true},
+        {ov::op::util::ScatterBase::get_type_info_static().name, true},
+        {ov::op::util::ScatterNDBase::get_type_info_static().name, true},
+        {ov::op::util::ScatterElementsUpdateBase::get_type_info_static().name, true},
+        {ov::op::v8::Slice::get_type_info_static().name, true},
+        {ov::op::v1::Broadcast::get_type_info_static().name, true},
+        {ov::op::v0::Concat::get_type_info_static().name, true},
+        {ov::op::v1::Split::get_type_info_static().name, true},
+        {ov::op::v1::StridedSlice::get_type_info_static().name, true},
+        {ov::op::v0::Tile::get_type_info_static().name, true},
+        {ov::op::v16::Identity::get_type_info_static().name, true},
+        {ov::op::v1::Pad::get_type_info_static().name, true},
+        {ov::op::util::GatherNDBase::get_type_info_static().name, true},
+        {ov::op::util::GatherBase::get_type_info_static().name, true},
+        {ov::op::v15::Squeeze::get_type_info_static().name, true},
+        {ov::op::v0::Unsqueeze::get_type_info_static().name, true},
+        // Add more node types as needed
+    };
+
+    auto it = node_behavior_registry.find(node->get_type_info().name);
+    return it != node_behavior_registry.end() && it->second;
+}
+
+ov::intel_gpu::DisableFP16CompressionForPeriodicFuncs::DisableFP16CompressionForPeriodicFuncs()
+    : ov::pass::MatcherPass() {
+    set_name("DisableFP16CompressionForPeriodicFuncs");
+
+    auto cos = ov::pass::pattern::wrap_type<ov::op::v0::Cos>();
+    auto sin = ov::pass::pattern::wrap_type<ov::op::v0::Sin>();
+    auto sin_cos_func_pattern = std::make_shared<ov::pass::pattern::op::Or>(OutputVector{cos, sin});
+
+    ov::matcher_pass_callback callback = [&](ov::pass::pattern::Matcher& m) {
+        auto node = m.get_match_root();
+
+        if (node->get_input_element_type(0) != ov::element::f16
+            && node->get_input_element_type(0) != ov::element::f32) {
+            return false;
+        }
+
+        // Disable FP16 compression for the current node
+        ov::disable_fp16_compression(node);
+
+        // Traverse inputs to find the first node that modifies values
+        std::function<bool(const std::shared_ptr<ov::Node>&)> traverse_inputs = [&](const std::shared_ptr<ov::Node>& current_node) -> bool {
```

Comment:
nit: `traverse_inputs` is too general name. What about `propagate_disabling_fp16_compression`?

Reply:
fixed it.